docs/proposals/device-crd.md

```diff
@@ -0,0 +1,671 @@
+    modbus:
+      register: CoilRegister
+      offset: 2
+      limit: 1
+      scale: 1.0
+      isSwap: true
+      isRegisterSwap: true
+  - propertyName: temperature-enable
+    modbus:
+      register: DiscreteInputRegister
+      offset: 3
+      limit: 1
+      scale: 1.0
+      isSwap: true
+      isRegisterSwap: true
+```
+Shown above is an example device model for a temperature sensor with Modbus protocol. It has two properties:
+- `temperature`: the temperature readings from the sensor. The `type` field indicates that the temperature property is of type `int`, it is read-only and the maximum value it can take is 100.
+- `temperature-enable`: this property defines whether data collection is enabled from the sensor. It is a writable property. Data collection is disabled if it is set to `OFF`. To turn on data collection, it needs to be set to `ON`.
+
+Property visitors provide details like how to access device properties. In the above example, there are two visitors defined which describe how to read/write the device properties using `modbus` protocol. Detailed information on the Modbus registers to access is provided along with the offset, limit and other settings.
+
+### Device instance CRD
+<img src="../images/device-crd/device-crd.png">
+
+A `device` instance represents an actual device object. It is like an instantiation of the `device model` and references properties defined in the model. The device spec is static while the device status contains dynamically changing data like the desired state of a device property and the state reported by the device.
+
+### Device instance type definition
+```go
+// DeviceSpec represents the static information of a single device instance.
+type DeviceSpec struct {
+	// Required: DeviceModelRef is reference to the device model used as a template
+	// to create the device instance.
+	DeviceModelRef *core.LocalObjectReference `json:"deviceModelRef,omitempty"`
+	// Required: The protocol configuration used to connect to the device.
+	Protocol       ProtocolConfig             `json:"protocol,omitempty"`
+	// NodeSelector indicates the binding preferences between devices and nodes.
+	// Refer to k8s.io/kubernetes/pkg/apis/core NodeSelector for more details
+	// +optional
+	NodeSelector   *core.NodeSelector         `json:"nodeSelector,omitempty"`
+}
+
+// Only one of its members may be specified.
+type ProtocolConfig struct {
+	// Protocol configuration for opc-ua
+	// +optional
+	OpcUA  *ProtocolConfigOpcUA  `json:"opcua,omitempty"`
+	// Protocol configuration for modbus
+	// +optional
+	Modbus *ProtocolConfigModbus `json:"modbus,omitempty"`
+}
+
+type ProtocolConfigOpcUA struct {
+	// Required: The URL for opc server endpoint.
+	Url            string `json:"url,omitempty"`
+	// Username for access opc server.
+	// +optional
+	UserName       string `json:"userName,omitempty"`
+	// Password for access opc server.
+	// +optional
+	Password       string `json:"password,omitempty"`
```

Comment:
Is it reasonable to give password in the config? If we have to, should we encode it with base64 or other cryptographic algorithm? @zhenghaibo1008

Reply:
Just have a chat with @zhenghaibo1008 offline, we will store users' credentials with some cryptographic algorithms in cloud, so the credentials only visible for users.

I buy it :)